src/main/java/com/atlassian/db/replica/internal/ReplicaStatement.java

```diff
@@ -554,41 +561,35 @@ void recordWriteAfterQueryExecution() throws SQLException {
 
     public Statement getReadStatement(RouteDecisionBuilder decisionBuilder) {
         if (connectionProvider.getState().equals(MAIN)) {
-            decisionBuilder.reason(MAIN_CONNECTION_REUSE);
+            decisionBuilder.reason(MAIN_CONNECTION_REUSE);//.isWrite(true) ? ;
+            // sth was readonly, but need to happen on Main because of inconsistency
+            // I think we don't have info whether this is read  / write
             connectionProvider.getStateDecision().ifPresent(decisionBuilder::cause);
+
+
             return prepareWriteStatement(decisionBuilder);
         }
         final String sql = decisionBuilder.getSql();
-        isWriteOperation = sqlFunction.isFunctionCall(sql) || isUpdate(sql) || isDelete(sql);
+        isWriteOperation = sqlOpeartionType.isWriteOpearation(sql);
         if (isWriteOperation) {
-            decisionBuilder.reason(WRITE_OPERATION);
+            decisionBuilder.reason(WRITE_OPERATION).isWrite(true);
             return prepareWriteStatement(decisionBuilder);
         }
 
-        if (isSelectForUpdate(sql)) {
-            decisionBuilder.reason(LOCK);
+        if (sqlOpeartionType.isSelectForUpdate(sql)) {
+            decisionBuilder.reason(LOCK).isWrite(true);
             return prepareWriteStatement(decisionBuilder);
         }
 
+        // todo what should happen here UNKNOWN SQL operation type ?
```

Comment:
It shouldn't be unknown. The `sql` should be checked

Reply:
It shouldn't. 
But it might be, if SQL operation type is different thing than isWriteOperation.
What do you mean by 
> sql should be checked

Should it be parsed?

Reply:
So it was checked already with `isWriteOpearation` and `isSelectForUpdate`. What other possible write types are there?
As @wyrzyk [said](https://github.com/atlassian-labs/db-replica/pull/48#discussion_r585359230), the library knows. The code is just arranged in a way that's hard to extract that knowledge. We can refactor internals as much as we need. Incrementality is less important than feature power.

Reply:
Ok, I don't get yet how to extract that knowledge and I'm asking for any suggestion related to this. I've shared my try to add isWrite information. to `RouteDecision`
But first we're unsure about final definition of isWrite

https://github.com/atlassian-labs/db-replica/pull/49#issuecomment-788925839
https://github.com/atlassian-labs/db-replica/pull/49#issuecomment-788778672